# Ujian Akhir Semester Analitika Media Sosial

## Nama : Zahransyah Alfhat
## NIM : A12.2020.06396
## Kelas : 6501

## Recomender System dengan pemodelan filtering
filtering merupakan metode yang biasanya digunakan dalam sistem rekomendasi. Metode ini menggunakan informasi tentang pengguna atau item yang tersedia untuk memfilter dan mengurutkan rekomendasi yang akan diberikan kepada pengguna. Ada dua jenis filtering yang umum digunakan: Filtering Berbasis Pengguna: Metode ini menggunakan informasi tentang preferensi atau sejarah interaksi pengguna dengan item-item tertentu untuk memberikan rekomendasi yang sesuai dengan minat dan kebiasaan pengguna.
Filtering Berbasis Item: Metode ini menggunakan informasi tentang karakteristik atau fitur dari setiap item untuk mencocokkan dengan preferensi pengguna dan memberikan rekomendasi yang sesuai.
Kedua metode ini biasanya digunakan bersama-sama untuk memberikan rekomendasi yang lebih akurat kepada pengguna. Misalnya, sebuah sistem rekomendasi dapat menggunakan informasi tentang sejarah pembelian pengguna dan fitur produk untuk memberikan rekomendasi produk yang sesuai dengan minat dan kebiasaan belanja pengguna.

### pada kesempatan kali ini saya akan melakukan rekomendasi film jepang dengan menggunakan pandas and numpy
pandas merupakan library dalam phyton untuk memanipulasi dan menganalisis data secara efisien dalam menangani data terstruktur dan tidak terstruktur seperti yang ada pada database, file csv, maupun data yang di-srape dari web
sedangkan numpy digunakan dalam komputasi numerik. library ini memberikan kemampuan pengolahan masalah matematika maupun numerik secara efisien.
### pertama tama yang harus dilakukan adalah melakukan import library
seperti yang dijelaskan diatas, library yang digunakan adalah pandas dan numpy

In [2]:
import numpy as np
import pandas as pd

### kemudian langkah selanjutnya saya akan melakukan import data
data yang saya digunakan berbentuk CSV (CSV adalah singkatan dari Comma Separated Values, yang merupakan format file yang digunakan untuk menyimpan data tabular dengan menggunakan koma sebagai pemisah antar nilai. Data tabular ini biasanya terdiri dari baris dan kolom, dengan setiap baris mewakili satu record atau entri data, dan setiap kolom mewakili sebuah field atau atribut dari data tersebut). untuk datasetnya sendiri saya menggunakan Anime Dataset 2022 dengan #rank rank of the Anime Given by Anime Planet Website terdiri dari 18495 data, 17 colums yang terdiri dari (Rank, Name, Japanese_name, Type, Episodes, Studio, Release_season, Tags, Rating, Release_year, End_year, Description, Content_Warning, Related_Mange, Related_Anime, Voice_actord, Staff).

In [3]:
data_m = pd.read_csv("Anime.csv")

### kemudian saya akan menampikan informasi berkaitan dengan data yang telah saya import

In [4]:
data_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18495 entries, 0 to 18494
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Rank             18495 non-null  int64  
 1   Name             18495 non-null  object 
 2   Japanese_name    7938 non-null   object 
 3   Type             18495 non-null  object 
 4   Episodes         9501 non-null   float64
 5   Studio           12018 non-null  object 
 6   Release_season   4116 non-null   object 
 7   Tags             18095 non-null  object 
 8   Rating           15364 non-null  float64
 9   Release_year     18112 non-null  float64
 10  End_year         2854 non-null   float64
 11  Description      18491 non-null  object 
 12  Content_Warning  1840 non-null   object 
 13  Related_Mange    7627 non-null   object 
 14  Related_anime    10063 non-null  object 
 15  Voice_actors     15309 non-null  object 
 16  staff            13005 non-null  object 
dtypes: float64(4

In [5]:
data_m.head()

Rank                                               Name  \
0     1  Demon Slayer: Kimetsu no Yaiba - Entertainment...   
1     2                     Fruits Basket the Final Season   
2     3                                    Mo Dao Zu Shi 3   
3     4                   Fullmetal Alchemist: Brotherhood   
4     5                Attack on Titan 3rd Season: Part II   

                                     Japanese_name   Type  Episodes  \
0                    Kimetsu no Yaiba: Yuukaku-hen  TV          NaN   
1                          Fruits Basket the Final  TV         13.0   
2                       The Founder of Diabolism 3  Web        12.0   
3   Hagane no Renkinjutsushi: Full Metal Alchemist  TV         64.0   
4             Shingeki no Kyojin Season 3: Part II  TV         10.0   

              Studio Release_season  \
0           ufotable          Fall    
1  TMS Entertainment         Spring   
2   B.C MAY PICTURES            NaN   
3              Bones         Spring   
4         WIT Studio         Spring   

                                                Tags  Rating  Release_year  \
0  Action, Adventure, Fantasy, Shounen, Demons, H...    4.60        2021.0   
1  Drama, Fantasy, Romance, Shoujo, Animal Transf...    4.60        2021.0   
2  Fantasy, Ancient China, Chinese Animation, Cul...    4.58        2021.0   
3  Action, Adventure, Drama, Fantasy, Mystery, Sh...    4.58        2009.0   
4  Action, Fantasy, Horror, Shounen, Dark Fantasy...    4.57        2019.0   

   End_year                                        Description  \
0       NaN  'Tanjiro and his friends accompany the Hashira...   
1       NaN                  'The final arc of Fruits Basket.'   
2       NaN               'The third season of Mo Dao Zu Shi.'   
3    2010.0  "The foundation of alchemy is based on the law...   
4       NaN  'The battle to retake Wall Maria begins now! W...   

                                     Content_Warning  \
0                                  Explicit Violence   
1  Emotional Abuse,, Mature Themes,, Physical Abu...   
2                                                NaN   
3  Animal Abuse,, Mature Themes,, Violence,, Dome...   
4                    Cannibalism,, Explicit Violence   

                                       Related_Mange  \
0                     Demon Slayer: Kimetsu no Yaiba   
1               Fruits Basket, Fruits Basket Another   
2  Grandmaster of Demonic Cultivation: Mo Dao Zu ...   
3  Fullmetal Alchemist, Fullmetal Alchemist (Ligh...   
4  Attack on Titan, Attack on Titan: End of the W...   

                                       Related_anime  \
0  Demon Slayer: Kimetsu no Yaiba, Demon Slayer: ...   
1  Fruits Basket 1st Season, Fruits Basket 2nd Se...   
2                   Mo Dao Zu Shi 2, Mo Dao Zu Shi Q   
3  Fullmetal Alchemist: Brotherhood Specials, Ful...   
4  Attack on Titan, Attack on Titan 2nd Season, A...   

                                        Voice_actors  \
0  Inosuke Hashibira : Yoshitsugu Matsuoka, Nezuk...   
1  Akito Sohma : Maaya Sakamoto, Kyo Sohma : Yuum...   
2  Lan Wangji, Wei Wuxian, Jiang Cheng, Jin Guang...   
3  Alphonse Elric : Rie Kugimiya, Edward Elric : ...   
4  Armin Arlelt : Marina Inoue, Eren Jaeger : Yuu...   

                                               staff  
0  Koyoharu Gotouge : Original Creator, Haruo Sot...  
1  Natsuki Takaya : Original Creator, Yoshihide I...  
2  Mo Xiang Tong Xiu : Original Creator, Xiong Ke...  
3  Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4  Hajime Isayama : Original Creator, Tetsurou Ar...

In [6]:
data_m.tail()

Rank                                               Name Japanese_name  \
18490  18491  Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...           NaN   
18491  18492                      Yi Tang Juchang: Sanguo Yanyi           NaN   
18492  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu           NaN   
18493  18494                      Chengshi Jiyi Wo Men de Jieri           NaN   
18494  18495  Heisei Inu Monogatari Bow: Genshi Inu Monogata...           NaN   

        Type  Episodes                        Studio Release_season  \
18490  Web         2.0  Sparkly Key Animation Studio            NaN   
18491  TV        108.0                           NaN            NaN   
18492  TV         13.0                           NaN            NaN   
18493  TV          NaN                           NaN            NaN   
18494  Movie       NaN              Nippon Animation            NaN   

                                                    Tags  Rating  \
18490  Action, Ancient China, Chinese Animation, Hist...     NaN   
18491                                  Chinese Animation     NaN   
18492  Chinese Animation, Family Friendly, Short Epis...     NaN   
18493  Chinese Animation, Family Friendly, Short Epis...     NaN   
18494                        Comedy, Slice of Life, Dogs     NaN   

       Release_year  End_year  \
18490        2020.0       NaN   
18491        2010.0       NaN   
18492        2020.0       NaN   
18493        2020.0       NaN   
18494        1994.0       NaN   

                                             Description Content_Warning  \
18490  Special episodes of Qin Shi Mingyue: Canghai H...             NaN   
18491                 No synopsis yet - check back soon!             NaN   
18492                 No synopsis yet - check back soon!             NaN   
18493                 No synopsis yet - check back soon!             NaN   
18494                 No synopsis yet - check back soon!             NaN   

      Related_Mange                     Related_anime Voice_actors staff  
18490           NaN  Qin Shi Mingyue: Canghai Hengliu          NaN   NaN  
18491           NaN                               NaN          NaN   NaN  
18492           NaN                               NaN          NaN   NaN  
18493           NaN                               NaN          NaN   NaN  
18494           NaN         Heisei Inu Monogatari Bow          NaN   NaN

## selanjutnya saya akan menampilkan isi dari beberapa column

In [8]:
data_m["Tags"][5]

'Action, Horror, Shounen, Curse, Exorcists, Monsters, School Life, Supernatural, Based on a Manga, Explicit Violence'

In [9]:
data_m["Description"][5]

"'Although Yuji Itadori looks like your average teenager, his immense physical strength is something to behold! Every sports club wants him to join, but Itadori would rather hang out with the school outcasts in the Occult Research Club. One day, the club manages to get their hands on a sealed cursed object. Little do they know the terror they’ll unleash when they break the seal…'"

In [10]:
data_m["Content_Warning"][5]

'Explicit Violence'

In [11]:
data_m["Related_Mange"][5]

'Jujutsu Kaisen 0, Jujutsu Kaisen'

In [12]:
data_m["staff"][5]

'Gege Akutami : Original Creator, Seong-Hu Park : Director, Arisa Okehazama : Music, Hiroaki Tsutsumi : Music, Yoshimasa Terui : Music, Tadashi Hiramatsu : Character Design, Akiko Fujita : Sound Director, Toshihiro Maeda : Producer'

### selanjutnya adalah melakukan filtering kolom untuk model
### nah disini fitur yang diperlukan adalah = (rank, name, type, tags, description, dan staff)

In [13]:
data_m = data_m[["Rank", "Name", "Type", "Tags", "Description", "staff"]]

### selanjutnya saya melakukan pengecekan apakah data tersebut terdapat duplikasi atau tidak
bisa kita lihat bahwa data duplikasi menunjukkan angka "0" sehingga kita bisa langsung masuk ke tahap selanjutnya

In [14]:
data_m.duplicated().sum()

0

### selanjutnya saya akan memeriksa nilai kosong yang ada pada data


In [15]:
data_m.isnull().sum()

Rank              0
Name              0
Type              0
Tags            400
Description       4
staff          5490
dtype: int64

### kemudian saya akan menampilkan nilai data yang ada pada atribut "type"

In [16]:
data_m["Type"].value_counts()

TV       5446
Movie    3577
Web      2488
OVA      2235
Music    2165
Other     990
DVD S     911
TV Sp     683
Name: Type, dtype: int64

## pada tahap ini saya akan melakukan cleaning pada data
cleaning data merupakan proses untuk menghilangkan kesalahan, kekurangan, dan inkonsisten data yang digunakan dalam menganalisis meliputi tindakan untuk menghilangkan duplikasi, mengisi data yang hilang, dan menyelaraskan data yang hilang. cleaning data sangat penting karena data yang tidak bersih dapat menghasilkan data yang tidak akurat
## variabel type dengan menghilangkan atribute mana saja yang sekiranya terdapat nilai kosong didalamnya

In [18]:
data_m[data_m["Type"] == "TV   "]

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...  TV      
1          2                     Fruits Basket the Final Season  TV      
3          4                   Fullmetal Alchemist: Brotherhood  TV      
4          5                Attack on Titan 3rd Season: Part II  TV      
5          6                                     Jujutsu Kaisen  TV      
...      ...                                                ...    ...   
18481  18482                                    Goblin Slayer 2  TV      
18488  18489        Shachiku-san wa Youjo Yuurei ni Iyasaretai.  TV      
18491  18492                      Yi Tang Juchang: Sanguo Yanyi  TV      
18492  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu  TV      
18493  18494                      Chengshi Jiyi Wo Men de Jieri  TV      

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
5      Action, Horror, Shounen, Curse, Exorcists, Mon...   
...                                                  ...   
18481  Action, Adventure, Fantasy, Seinen, Dark Fanta...   
18488  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   
18491                                  Chinese Animation   
18492  Chinese Animation, Family Friendly, Short Epis...   
18493  Chinese Animation, Family Friendly, Short Epis...   

                                             Description  \
0      'Tanjiro and his friends accompany the Hashira...   
1                      'The final arc of Fruits Basket.'   
3      "The foundation of alchemy is based on the law...   
4      'The battle to retake Wall Maria begins now! W...   
5      'Although Yuji Itadori looks like your average...   
...                                                  ...   
18481  It's fall, and the village's harvest festival ...   
18488  The story follows the daily life of Fushihara,...   
18491                 No synopsis yet - check back soon!   
18492                 No synopsis yet - check back soon!   
18493                 No synopsis yet - check back soon!   

                                                   staff  
0      Koyoharu Gotouge : Original Creator, Haruo Sot...  
1      Natsuki Takaya : Original Creator, Yoshihide I...  
3      Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4      Hajime Isayama : Original Creator, Tetsurou Ar...  
5      Gege Akutami : Original Creator, Seong-Hu Park...  
...                                                  ...  
18481                      Kumo Kagyu : Original Creator  
18488                     Imari Arita : Original Creator  
18491                                                NaN  
18492                                                NaN  
18493                                                NaN  

[5446 rows x 6 columns]

### selanjutnya saya akan melakukan filter terhadap kolom "type" dengan menghapus NaN yang ada dan memasukkannya kembali ke dalam dataframe
NaN adalah singkatan dari Not a Number, yang merupakan sebuah nilai yang digunakan untuk menunjukkan bahwa suatu operasi matematika tidak memiliki hasil yang valid. Nilai ini biasanya muncul dalam komputasi numerik ketika terjadi kesalahan atau tidak dapat menghasilkan hasil yang valid.

In [19]:
temp = data_m["Type"].apply(lambda x: [" ".join(x.split()) for i in x])

In [20]:
def type_(input):
    L = []
    for i in input:
        L.append(i[0])
    return L

In [21]:
data_m["Type"]= pd.Series(type_(temp))

### disini atribut tak bernilai yang tidak diperlukan telah hilang

In [22]:
data_m[data_m["Type"] == "TV"]

Rank                                               Name Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...   TV   
1          2                     Fruits Basket the Final Season   TV   
3          4                   Fullmetal Alchemist: Brotherhood   TV   
4          5                Attack on Titan 3rd Season: Part II   TV   
5          6                                     Jujutsu Kaisen   TV   
...      ...                                                ...  ...   
18481  18482                                    Goblin Slayer 2   TV   
18488  18489        Shachiku-san wa Youjo Yuurei ni Iyasaretai.   TV   
18491  18492                      Yi Tang Juchang: Sanguo Yanyi   TV   
18492  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu   TV   
18493  18494                      Chengshi Jiyi Wo Men de Jieri   TV   

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
5      Action, Horror, Shounen, Curse, Exorcists, Mon...   
...                                                  ...   
18481  Action, Adventure, Fantasy, Seinen, Dark Fanta...   
18488  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   
18491                                  Chinese Animation   
18492  Chinese Animation, Family Friendly, Short Epis...   
18493  Chinese Animation, Family Friendly, Short Epis...   

                                             Description  \
0      'Tanjiro and his friends accompany the Hashira...   
1                      'The final arc of Fruits Basket.'   
3      "The foundation of alchemy is based on the law...   
4      'The battle to retake Wall Maria begins now! W...   
5      'Although Yuji Itadori looks like your average...   
...                                                  ...   
18481  It's fall, and the village's harvest festival ...   
18488  The story follows the daily life of Fushihara,...   
18491                 No synopsis yet - check back soon!   
18492                 No synopsis yet - check back soon!   
18493                 No synopsis yet - check back soon!   

                                                   staff  
0      Koyoharu Gotouge : Original Creator, Haruo Sot...  
1      Natsuki Takaya : Original Creator, Yoshihide I...  
3      Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4      Hajime Isayama : Original Creator, Tetsurou Ar...  
5      Gege Akutami : Original Creator, Seong-Hu Park...  
...                                                  ...  
18481                      Kumo Kagyu : Original Creator  
18488                     Imari Arita : Original Creator  
18491                                                NaN  
18492                                                NaN  
18493                                                NaN  

[5446 rows x 6 columns]

### dilanjutkan untuk melakukan observasi tanpa tags dan description yang sekiranya tidak perlu digunakan

In [23]:
filtering_1 = data_m[ (data_m["Description"] == "'No synopsis yet - check back soon!'") & (data_m["Tags"].isnull())].index
data_m = data_m.drop(filtering_1)

In [24]:
data_m.isnull().sum()

Rank              0
Name              0
Type              0
Tags            267
Description       4
staff          5451
dtype: int64

### selanjutnya saya akan mencoba untuk melakukan drop pada tipe music, ova, tv sp. hal ini dikarenakan kebanyakan merupakan anime filler

In [25]:
filtering_2 = data_m[ (data_m["Type"] == "Music") | (data_m["Type"] == "OVA") | (data_m["Type"] == "TV Sp")].index
data_m.drop(filtering_2,inplace = True)

In [26]:
data_m.isnull().sum()

Rank              0
Name              0
Type              0
Tags             31
Description       4
staff          4322
dtype: int64

In [27]:
filtering_3 = data_m[ (data_m["Description"] == "No synopsis yet - check back soon!") & (data_m["Tags"].isnull())].index
data_m.drop(filtering_3,inplace=True)

In [28]:
data_m[data_m["Tags"].isnull()]

Rank                                               Name   Type Tags  \
7910    7911                         Himitsu no Akko-chan Movie  Movie  NaN   
8325    8326  Himitsu no Akko-chan: Umi da! Obake da!! Natsu...  Movie  NaN   
8341    8342                        Nanaka 6/17: Ojamana Nanaka  DVD S  NaN   
10185  10186                               Tobe! Kujira no Peek  Movie  NaN   

                                             Description  \
7910   "Taisho's father plans to construct an apartme...   
8325   "Akko spends a summer with her friends at her ...   
8341   'Unaired episode included in the DVD release o...   
10185  'The story of an albino whale used in a circus...   

                                                   staff  
7910   Fujio Akatsuka : Original Creator, Hiroki Shib...  
8325   Fujio Akatsuka : Original Creator, Hiroki Shib...  
8341                                                 NaN  
10185  Kouji Morimoto : Director, Satoru Utsunomiya :...

## pada kali ini atribute "tags" memiliki peran yang penting dalam pemilihan rekomendasi anime dengan observasi rank 7911, 8325, 8342, 10186 akan dihapus

In [29]:
filtering_4 = data_m[data_m["Tags"].isnull()].index
data_m.drop(filtering_4,inplace=True)

In [30]:
data_m.isnull().sum()

Rank              0
Name              0
Type              0
Tags              0
Description       4
staff          4300
dtype: int64

In [31]:
data_m.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13356 entries, 0 to 18494
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Rank         13356 non-null  int64 
 1   Name         13356 non-null  object
 2   Type         13356 non-null  object
 3   Tags         13356 non-null  object
 4   Description  13352 non-null  object
 5   staff        9056 non-null   object
dtypes: int64(1), object(5)
memory usage: 730.4+ KB


### pada description column, kita bisa melihat bahwa tidak semua data memiliki description. maka dari itu dilakukan drop

In [32]:
data_m.drop("Description", axis=1, inplace=True)

In [33]:
data_m.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13356 entries, 0 to 18494
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Rank    13356 non-null  int64 
 1   Name    13356 non-null  object
 2   Type    13356 non-null  object
 3   Tags    13356 non-null  object
 4   staff   9056 non-null   object
dtypes: int64(1), object(4)
memory usage: 626.1+ KB


### setelah mengamati variabel staff, saya menemukan terdapat 5490 missing value. maka dari itu kita hanya akan mengambil observasi dengan variabel staff yang berisikan nama asli penciptanya

In [34]:
data_m["staff"].isna().sum()

4300

### selanjutnya akan dilakukan filtering terhadap variabel staff

In [35]:

data_m_mini = data_m[data_m["staff"].notnull()]

In [36]:
data_m_mini.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9056 entries, 0 to 18489
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Rank    9056 non-null   int64 
 1   Name    9056 non-null   object
 2   Type    9056 non-null   object
 3   Tags    9056 non-null   object
 4   staff   9056 non-null   object
dtypes: int64(1), object(4)
memory usage: 424.5+ KB


In [37]:
data_m_mini["staff"][0]

'Koyoharu Gotouge : Original Creator, Haruo Sotozaki : Director, Akira Matsushima : Character Design, Aimer : Song Performance'

In [38]:
data_m_mini["staff"][1]

'Natsuki Takaya : Original Creator, Yoshihide Ibata : Director & Episode Director & Storyboard, Taku Kishimoto : Screenplay & Series Composition, Masaru Yokoyama : Music, Masaru Shindou : Character Design & Chief Animation Director, Baek-Ryun Chae : Photography Director, Youko Koyama : Art Director, Mika Sugawara : Color Design'

In [39]:
data_m_mini

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1          2                     Fruits Basket the Final Season     TV   
2          3                                    Mo Dao Zu Shi 3    Web   
3          4                   Fullmetal Alchemist: Brotherhood     TV   
4          5                Attack on Titan 3rd Season: Part II     TV   
...      ...                                                ...    ...   
18484  18485                                           Tooriame  Movie   
18486  18487                                   Better back then  Movie   
18487  18488                                        Make My Day    Web   
18488  18489        Shachiku-san wa Youjo Yuurei ni Iyasaretai.     TV   
18489  18490                                          Nostalgia  Movie   

                                                    Tags  \
0      Action, Adventure, Fantasy, Shounen, Demons, H...   
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...   
2      Fantasy, Ancient China, Chinese Animation, Cul...   
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...   
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...   
...                                                  ...   
18484                                   Abstract, Shorts   
18486       Shorts, Stop Motion Animation, Original Work   
18487                    Horror, Monsters, Original Work   
18488  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...   
18489  Drama, No Dialogue, Shorts, Stop Motion Animation   

                                                   staff  
0      Koyoharu Gotouge : Original Creator, Haruo Sot...  
1      Natsuki Takaya : Original Creator, Yoshihide I...  
2      Mo Xiang Tong Xiu : Original Creator, Xiong Ke...  
3      Hiromu Arakawa : Original Creator, Yasuhiro Ir...  
4      Hajime Isayama : Original Creator, Tetsurou Ar...  
...                                                  ...  
18484  Tomoko Takayama : Director, Juri Ito : Music, ...  
18486  Naoaki Shibuta : Director, Chikara Uemizutaru ...  
18487  Yasuo Ootagaki : Original Creator, Makoto Kane...  
18488                     Imari Arita : Original Creator  
18489                         Tomoyasu Murata : Director  

[9056 rows x 5 columns]

### langkah selanjutnya adalah melakukan reset index untuk melakukan konversi terhadap staff dari series str ke dict
dict adalah tipe data di Python yang merupakan sebuah tipe data koleksi (collection) yang terdiri dari pasangan key-value. Key merupakan identifier unik dari setiap value.

Sedangkan str adalah tipe data di Python yang merupakan representasi dari string atau teks.

In [40]:
data_m_mini.reset_index(inplace = True)

### convert ke dict

In [41]:
try_1 = data_m_mini["staff"][1].split(",")
try_1 = {k.split(":")[0].strip() : k.split(":")[1].strip() for k in try_1}
print(try_1)

{'Natsuki Takaya': 'Original Creator', 'Yoshihide Ibata': 'Director & Episode Director & Storyboard', 'Taku Kishimoto': 'Screenplay & Series Composition', 'Masaru Yokoyama': 'Music', 'Masaru Shindou': 'Character Design & Chief Animation Director', 'Baek-Ryun Chae': 'Photography Director', 'Youko Koyama': 'Art Director', 'Mika Sugawara': 'Color Design'}


### series str ke dict

In [42]:
for i in range (0,len(data_m_mini["staff"])):
    try_1 = data_m_mini["staff"][i].split(",")
    data_m_mini["staff"][i] = {k.split(":")[0].strip() : k.split(":")[1].strip() for k in try_1}

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\793470206.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini["staff"][i] = {k.split(":")[0].strip() : k.split(":")[1].strip() for k in try_1}


In [43]:
data_m_mini["staff"][1]

{'Natsuki Takaya': 'Original Creator',
 'Yoshihide Ibata': 'Director & Episode Director & Storyboard',
 'Taku Kishimoto': 'Screenplay & Series Composition',
 'Masaru Yokoyama': 'Music',
 'Masaru Shindou': 'Character Design & Chief Animation Director',
 'Baek-Ryun Chae': 'Photography Director',
 'Youko Koyama': 'Art Director',
 'Mika Sugawara': 'Color Design'}

In [44]:
data_m_mini["staff"][1].keys()

dict_keys(['Natsuki Takaya', 'Yoshihide Ibata', 'Taku Kishimoto', 'Masaru Yokoyama', 'Masaru Shindou', 'Baek-Ryun Chae', 'Youko Koyama', 'Mika Sugawara'])

In [45]:
data_m_mini["staff"][1].values()

dict_values(['Original Creator', 'Director & Episode Director & Storyboard', 'Screenplay & Series Composition', 'Music', 'Character Design & Chief Animation Director', 'Photography Director', 'Art Director', 'Color Design'])

### selanjutnya yaitu mengkonversi values ke keys dan keys ke value pada dict

In [46]:
for i in range (0,len(data_m_mini["staff"])):
    dict = data_m_mini["staff"][i]
    data_m_mini["staff"][i] = {value:key for key, value in dict.items()}

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\1896485708.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini["staff"][i] = {value:key for key, value in dict.items()}


In [47]:
data_m_mini.iloc[8000:8023,3:4]

Type
8000  Movie
8001     TV
8002  Movie
8003     TV
8004    Web
8005    Web
8006     TV
8007  Movie
8008     TV
8009     TV
8010  Movie
8011     TV
8012  Other
8013  Movie
8014  Movie
8015    Web
8016     TV
8017  Movie
8018     TV
8019     TV
8020    Web
8021    Web
8022    Web

### selanjutnya adalah melakukan filtering terhadap original creator pada kolom

In [48]:
for i in range (0,len(data_m_mini["staff"])):
    data_m_mini["staff"][i] = { key:value for (key,value) in data_m_mini["staff"][i].items() if key  == "Original Creator"}

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\3047246284.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini["staff"][i] = { key:value for (key,value) in data_m_mini["staff"][i].items() if key  == "Original Creator"}


In [49]:
data_m_mini.iloc[1:40,3:4]

Type
1      TV
2     Web
3      TV
4      TV
5      TV
6      TV
7      TV
8   Movie
9      TV
10  Movie
11     TV
12     TV
13     TV
14  Movie
15  Movie
16     TV
17    Web
18     TV
19     TV
20     TV
21    Web
22  Movie
23    Web
24     TV
25     TV
26     TV
27     TV
28  Movie
29     TV
30    Web
31     TV
32     TV
33     TV
34     TV
35    Web
36     TV
37     TV
38    Web
39  Movie

### filtering data untuk dict dengan value

In [50]:
filtering_5 = data_m_mini[data_m_mini["staff"] == {}].index
data_m_mini.drop(filtering_5,inplace = True)

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\716305122.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini.drop(filtering_5,inplace = True)


### selanjutnya reset index

In [51]:
data_m_mini.reset_index(inplace = True)

### selanjutnya pertahankan value dari original creator

In [52]:
for i in range (0,len(data_m_mini["staff"])):
        data_m_mini["staff"][i] = (data_m_mini["staff"][i])["Original Creator"]

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\830981870.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini["staff"][i] = (data_m_mini["staff"][i])["Original Creator"]


## nah sekarang telah sukses mengekstrak nama dari original creator

In [53]:
data_m_mini

level_0  index   Rank  \
0           0      0      1   
1           1      1      2   
2           2      2      3   
3           3      3      4   
4           4      4      5   
...       ...    ...    ...   
3194     9037  18443  18444   
3195     9049  18481  18482   
3196     9050  18483  18484   
3197     9053  18487  18488   
3198     9054  18488  18489   

                                                   Name   Type  \
0     Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1                        Fruits Basket the Final Season     TV   
2                                       Mo Dao Zu Shi 3    Web   
3                      Fullmetal Alchemist: Brotherhood     TV   
4                   Attack on Titan 3rd Season: Part II     TV   
...                                                 ...    ...   
3194                                      The Imaginary  Movie   
3195                                    Goblin Slayer 2     TV   
3196                        Peleliu: Rakuen no Guernica  Other   
3197                                        Make My Day    Web   
3198        Shachiku-san wa Youjo Yuurei ni Iyasaretai.     TV   

                                                   Tags              staff  
0     Action, Adventure, Fantasy, Shounen, Demons, H...   Koyoharu Gotouge  
1     Drama, Fantasy, Romance, Shoujo, Animal Transf...     Natsuki Takaya  
2     Fantasy, Ancient China, Chinese Animation, Cul...  Mo Xiang Tong Xiu  
3     Action, Adventure, Drama, Fantasy, Mystery, Sh...     Hiromu Arakawa  
4     Action, Fantasy, Horror, Shounen, Dark Fantasy...     Hajime Isayama  
...                                                 ...                ...  
3194                                   Based on a Novel      A. F. Harrold  
3195  Action, Adventure, Fantasy, Seinen, Dark Fanta...         Kumo Kagyu  
3196  Action, Shounen, Island, War, World War 2, Bas...    Masao Hiratsuka  
3197                    Horror, Monsters, Original Work     Yasuo Ootagaki  
3198  Comedy, Slice of Life, Ghosts, Iyashikei, Non-...        Imari Arita  

[3199 rows x 7 columns]

In [54]:
data_m_mini["Tags"] = data_m_mini["Tags"] +", "+ data_m_mini["staff"]

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\2256362315.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini["Tags"] = data_m_mini["Tags"] +", "+ data_m_mini["staff"]


In [55]:
data_m_mini.drop(labels = "staff",axis = 1, inplace= True)


C:\Users\acer\AppData\Local\Temp\ipykernel_1608\2189139580.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m_mini.drop(labels = "staff",axis = 1, inplace= True)


In [56]:
data_m.drop("staff",axis = 1, inplace = True)

In [57]:
data_m["Tags"][0]

'Action, Adventure, Fantasy, Shounen, Demons, Historical, Martial Arts, Orphans, Siblings, Swordplay, Based on a Manga, Explicit Violence'

In [58]:
data_m_mini["Tags"][0]

'Action, Adventure, Fantasy, Shounen, Demons, Historical, Martial Arts, Orphans, Siblings, Swordplay, Based on a Manga, Explicit Violence, Koyoharu Gotouge'

In [59]:
data_m.set_index(['Rank','Name'], inplace=True)
data_m.update(data_m_mini.set_index(['Rank','Name']))
data_m.reset_index(inplace=True)
data_m

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1          2                     Fruits Basket the Final Season     TV   
2          3                                    Mo Dao Zu Shi 3    Web   
3          4                   Fullmetal Alchemist: Brotherhood     TV   
4          5                Attack on Titan 3rd Season: Part II     TV   
...      ...                                                ...    ...   
13351  18491  Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...    Web   
13352  18492                      Yi Tang Juchang: Sanguo Yanyi     TV   
13353  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu     TV   
13354  18494                      Chengshi Jiyi Wo Men de Jieri     TV   
13355  18495  Heisei Inu Monogatari Bow: Genshi Inu Monogata...  Movie   

                                                    Tags  
0      Action, Adventure, Fantasy, Shounen, Demons, H...  
1      Drama, Fantasy, Romance, Shoujo, Animal Transf...  
2      Fantasy, Ancient China, Chinese Animation, Cul...  
3      Action, Adventure, Drama, Fantasy, Mystery, Sh...  
4      Action, Fantasy, Horror, Shounen, Dark Fantasy...  
...                                                  ...  
13351  Action, Ancient China, Chinese Animation, Hist...  
13352                                  Chinese Animation  
13353  Chinese Animation, Family Friendly, Short Epis...  
13354  Chinese Animation, Family Friendly, Short Epis...  
13355                        Comedy, Slice of Life, Dogs  

[13356 rows x 4 columns]

In [60]:
for i in range(0,len(data_m["Tags"])):
    data_m["Tags"][i] = data_m["Tags"][i].replace(" ","")

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\2813507748.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m["Tags"][i] = data_m["Tags"][i].replace(" ","")


In [61]:
for i in range(0,len(data_m["Tags"])):
    data_m["Tags"][i] = data_m["Tags"][i].replace(","," ")

C:\Users\acer\AppData\Local\Temp\ipykernel_1608\2589152576.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_m["Tags"][i] = data_m["Tags"][i].replace(","," ")


In [62]:
data_m["Tags"] = data_m["Tags"].apply(lambda x: x.lower())

## Selanjutnya pembuatan model rekomendasi

In [63]:
data = data_m.copy()

In [64]:
data

Rank                                               Name   Type  \
0          1  Demon Slayer: Kimetsu no Yaiba - Entertainment...     TV   
1          2                     Fruits Basket the Final Season     TV   
2          3                                    Mo Dao Zu Shi 3    Web   
3          4                   Fullmetal Alchemist: Brotherhood     TV   
4          5                Attack on Titan 3rd Season: Part II     TV   
...      ...                                                ...    ...   
13351  18491  Qin Shi Mingyue: Canghai Hengliu Xiaomeng Spec...    Web   
13352  18492                      Yi Tang Juchang: Sanguo Yanyi     TV   
13353  18493          Fenghuang Ji Xiang Yu Qingming Shanghe Tu     TV   
13354  18494                      Chengshi Jiyi Wo Men de Jieri     TV   
13355  18495  Heisei Inu Monogatari Bow: Genshi Inu Monogata...  Movie   

                                                    Tags  
0      action adventure fantasy shounen demons histor...  
1      drama fantasy romance shoujo animaltransformat...  
2      fantasy ancientchina chineseanimation cultivat...  
3      action adventure drama fantasy mystery shounen...  
4      action fantasy horror shounen darkfantasy isol...  
...                                                  ...  
13351  action ancientchina chineseanimation historica...  
13352                                   chineseanimation  
13353      chineseanimation familyfriendly shortepisodes  
13354      chineseanimation familyfriendly shortepisodes  
13355                            comedy sliceoflife dogs  

[13356 rows x 4 columns]

## Bag Of words atau mengubah data teks menjadi vektor yang dapat dipahami oleh komputer.
Bag of Words (BoW) adalah sebuah metode yang digunakan dalam proses pengolahan teks (Natural Language Processing, NLP) untuk mengkonversi teks menjadi representasi numerik yang dapat diproses oleh mesin. Metode ini merupakan salah satu teknik dasar yang sering digunakan dalam NLP dan sangat berguna dalam berbagai aplikasi, seperti klasifikasi dokumen, analisis sentimen, dan pembuatan rekomendasi.

In [65]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 13500,stop_words="english")

In [66]:
cv.fit_transform(data["Tags"]).toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [67]:
cv.fit_transform(data["Tags"]).toarray().shape

(13356, 1931)

In [68]:
vectors = cv.fit_transform(data["Tags"]).toarray()

In [69]:
cv.get_feature_names()[0:10]

C:\Users\acer\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['15thcentury',
 '16thcentury',
 '17thcentury',
 '18thcentury',
 '19thcentury',
 'abiumeda',
 'abstract',
 'acomomota',
 'acting',
 'action']

In [70]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vectors)
similarity

array([[1.        , 0.19611614, 0.10482848, ..., 0.        , 0.        ,
        0.        ],
       [0.19611614, 1.        , 0.08908708, ..., 0.        , 0.        ,
        0.        ],
       [0.10482848, 0.08908708, 1.        , ..., 0.21821789, 0.21821789,
        0.        ],
       ...,
       [0.        , 0.        , 0.21821789, ..., 1.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.21821789, ..., 1.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [71]:
similarity.shape

(13356, 13356)

In [72]:
def Recommender(user_input):
    Index_of_anime = data[data["Name"] == user_input].index[0]
    Similarity_score = similarity[Index_of_anime]
    Sorted_scores = sorted(list(enumerate(Similarity_score)),reverse = True,key= lambda x: x[1]) [1:6]
    Recommended_Anime = []
    
    for i in Sorted_scores:
        Recommended_Anime.append(data.iloc[i[0]].Name)
    return Recommended_Anime

In [73]:
Recommender("Your lie in April")

['Orange',
 'The Flowers of Evil',
 'Forest of Piano 2nd Season',
 "Scum's Wish",
 'Kono Oto Tomare! 2nd Season']

In [74]:
Recommender("Yaoshenji 2nd Season")

['Yaoshenji 4th Season',
 'Yaoshenji 2nd Season',
 'Yaoshenji',
 'Yaoshenji 5th Season',
 'Star Martial God Technique']

In [75]:
Recommender("One Piece")

['One Piece Film: Z',
 'One Piece Film: Gold ~Episode 0~ 711 ver.',
 'One Piece Movie 10: Strong World',
 'One Piece Movie 1',
 'One Piece Movie 14: Stampede']

In [76]:
Recommender("My Hero Academia")

['My Hero Academia 3',
 'My Hero Academia 2',
 "My Hero Academia Movie 3: World Heroes' Mission",
 'My Hero Academia: Jump Festa 2016 Special',
 'My Hero Academia Movie 2: Heroes:Rising']

In [77]:
Recommender("Jujutsu Kaisen")

['Jujutsu Kaisen 0',
 'Ga-Rei-Zero',
 "Hell's Paradise: Jigokuraku PV",
 'Corpse Princess: Kuro',
 "JoJo's Bizarre Adventure: Phantom Blood"]

kesimpulannya adalah kita telah mendapat rekomendasi film yang kita inginkan berdasarkan judul film yang kita masukkan. data tersebut similar berdasarkan variabel yang diinputkan